lib/pbench/cli/agent/commands/tools/base.py

```diff
@@ -15,7 +15,7 @@ def tools(self, path):
         """List all tools in a given path"""
         return sorted(
             [
-                p.name
+                p
```

Comment:
For instance, is this change going to break the code at `lib/pbench/cli/agent/commands/tools/list.py` line 67?  (That is, in order for the `--name` option to match, it will need to be a full path, now, right?  I think that's a change in behavior from the user's point of view.)

Reply:
There is one more use in the [clear command](https://github.com/distributed-system-analysis/pbench/blob/d6ef508df6bffab1812401ddcafde5fe212172c5/lib/pbench/cli/agent/commands/tools/clear.py#L56) but the `_clear_tools` [method](https://github.com/distributed-system-analysis/pbench/blob/d6ef508df6bffab1812401ddcafde5fe212172c5/lib/pbench/cli/agent/commands/tools/clear.py#L83) concatenates paths: before this change it would concatenate a path with a (simple) name; now it concatenates two paths. Both seem to work: concatenating two paths takes care of the common prefix.

But @webbnh is correct that, *within* `list`, there are more problems: those are not found by the (currently weak) functional tests because I don't check output. So the next step is to: a) fix the expected status for (some of) the functional tests and make sure that the command returns the appropriate error status and b) take the output of the command into account. This should catch the `--with-option` problems as well as the problem that @webbnh indicated (and any others that might arise). My plan is to check the output for all the "successful" cases (as indicated by a 0 return status: if it should fail and it does fail with non-zero status, I consider the test a pass and I won't check the output, but if it succeeds and should succeed, the output should be checked as well. And of course, if it fails with a traceback, the test should fail anyway.

Does that sound reasonable?

Reply:
I'm a fan of confirming that we meet end-user expectations....

Since these commands are supposed to produce fairly specific output in success cases, it would be good to check, when the tool reports success, that it produced acceptable output.  Correspondingly, when the tool reports error/failure, it would be good to check that it produced appropriate output (presumably on `stderr` and not on `stdout`) in those cases, as well.

Reply:
Actually, although it is true that `_clear_tools` does not have any problem clearing the tool specified by its path, it does have a problem reporting what it did. So I will submit a fix for that in the next commit, but I have conceptual problems with its specification (and with its implementation as well), so I'll open an issue for it and address it later.

Reply:
@webbnh: I'll add tests for the error text in those cases, but ATM the output (normal and error) needs work.  I am going to leave it to the last, fixing the tests first to require the "correct" output and after we are all happy with that, fixing the program to produce that output. For the time being, I'm more worried about the output text in the normal cases containing all the information that is expected (albeit in not quite final form).

Reply:
Sounds good.  

(You tagged me and concluded with, "Does that sound reasonable?", and I felt like a simple "yes" or "no" response lacked nuance.  :-) )

Reply:
I like nuance :-)